src/lib/transaction_snark/test/zkapps_examples/thing/thing.ml

```diff
@@ -0,0 +1,183 @@
+open Transaction_snark_tests.Util
+open Core_kernel
+open Mina_base
+open Signature_lib
+module Impl = Pickles.Impls.Step
+module Inner_curve = Snark_params.Tick.Inner_curve
+module Nat = Pickles_types.Nat
+module Local_state = Mina_state.Local_state
+module Zkapp_command_segment = Transaction_snark.Zkapp_command_segment
+
+(* create keypair *)
+let sk = Private_key.create ()
+
+let pk = Public_key.of_private_key_exn sk
+
+let pk_compressed = Public_key.compress pk
+
+(* create account id from token + pubkey *)
+
+let account_id = Account_id.create pk_compressed Token_id.default
+
+(* we want to create a circuit with a domain of size 2^16 *)
+let num_constraints = 1 lsl 16
+
+let () = Format.printf "num_constraints: %d@." num_constraints
+
+(* create a zkapp? but the rules are wrap circuits, mmm... *)
+
+let tag, _cache, _p_module, Pickles.Provers.[ prover ] =
+  Zkapps_examples.compile () ~cache:Cache_dir.cache ~auxiliary_typ:Impl.Typ.unit
+    ~branches:(module Nat.N1)
+    ~max_proofs_verified:(module Nat.N0)
+    ~name:"empty_update"
+    ~constraint_constants:
+      (Genesis_constants.Constraint_constants.to_snark_keys_header
+         constraint_constants )
+    ~choices:(fun ~self:_ ->
+      [ Zkapps_big_circuit.rule ~num_constraints pk_compressed ] )
+
+(* get verification key from memory *)
+
+let vk : Side_loaded_verification_key.t =
+  Pickles.Side_loaded.Verification_key.of_compiled tag
+
+(* I couldn't figure out how to get the domain size from vk here
+   let () =
+      let domain_size = vk.step.domain_size in
+      assert (domain_size = 16)
+*)
```

Comment:
I couldn't figure out how to get the domain size from vk here.
If I jump to `Side_loaded_verification_key.t` I only have access to an interface that doesn't allow me to do much. Maybe that interface is used with some manual type equality (e.g. `Intf with t = some_other_type`) but I'm not sure how to find out where this would be.

Reply:
The domain size that we care about comes from the rule, so it can vary for the same vk. You're better checking the field of the proof (`deferred_values.branch_data.log2_domain_size` or whatever it's called).

Reply:
sorry I'm not sure I understand what you mean here :o

Reply:
OK I think I understand it as, the vk is the wrap vk and so it doesn't tell us about the domain size of the app (which is in the step vk).

The step vk is either hardcoded in the circuit of the wrap vk, or something we have to pass as public input (or as a witness, and a hash of that and some other metadata is passed as public input instead)

so for some reason it's also in a proof? So I have to find out where the proofs are.

I look at what the prover returns when executed:

```ocaml
Async.Thread_safe.block_on_async_exn prover
```

and it looks like it's a value of type

```
((Account_update.t, Zkapp_command.Digest.Account_update.t,
   Zkapp_command.Digest.Forest.t)
  Zkapp_command.Call_forest.tree * unit
```

which doesn't seem to contain an actual proof. I'm a bit confused here, I'll continue to investigate